dev/com.ibm.ws.kernel.service/src/com/ibm/ws/common/crypto/CryptoUtils.java

```diff
@@ -0,0 +1,401 @@
+            zOSAndJAVA11orHigherChecked = true;
+            return iszOSAndJava11orHigher;
+        }
+    }
+
+    public static String getProvider() {
+        String provider = null;
+
+        if (fipsEnabled) {
+            //Do not check the provider available or not. Later on when we use the provider, the JDK will handle it.
+            if (isSemeruFips()) {
+                provider = OPENJCE_PLUS_FIPS_NAME;
+            } else {
+                provider = IBMJCE_PLUS_FIPS_NAME;
+            }
+        } else if (isZOSandRunningJava11orHigher() && isOpenJCEPlusAvailable()) {
+            provider = OPENJCE_PLUS_NAME;
+        } else if (isIBMJCEAvailable()) {
+            provider = IBMJCE_NAME;
+        }
+
+        if (TraceComponent.isAnyTracingEnabled() && tc.isDebugEnabled()) {
+            if (provider != null) {
+                Tr.debug(tc, "Provider configured is " + provider);
+            } else {
+                Tr.debug(tc, "Provider configured by JDK is " + (Security.getProviders() != null ? Security.getProviders()[0].getName() : "NULL"));
+            }
+        }
+        return provider;
+    }
+
+    public static final String MESSAGE_DIGEST_ALGORITHM = (fipsEnabled ? MESSAGE_DIGEST_ALGORITHM_SHA256 : MESSAGE_DIGEST_ALGORITHM_SHA);
+    /**
+     * List of supported Message Digest Algorithms.
+     */
+    private static final List<String> supportedMessageDigestAlgorithms = Arrays.asList(
+                                                                                       MESSAGE_DIGEST_ALGORITHM_SHA256,
+                                                                                       MESSAGE_DIGEST_ALGORITHM_SHA384,
+                                                                                       MESSAGE_DIGEST_ALGORITHM_SHA512);
+
+    public static String getMessageDigestAlgorithm() {
+        return MESSAGE_DIGEST_ALGORITHM_SHA256;
+    }
+
+    public static MessageDigest getMessageDigest() throws NoSuchAlgorithmException {
+        return getMessageDigest(getMessageDigestAlgorithm());
+    }
+
+    public static MessageDigest getMessageDigest(String algorithm) throws NoSuchAlgorithmException {
+        if (!supportedMessageDigestAlgorithms.contains(algorithm))
+            throw new NoSuchAlgorithmException(String.format("Algorithm %s is not supported", algorithm));
+        return MessageDigest.getInstance(algorithm);
+    }
+
+    public static MessageDigest getMessageDigestForLTPA() {
+        MessageDigest md1 = null;
+        try {
+            if (fipsEnabled) {
+                if (isSemeruFips()) {
+                    md1 = MessageDigest.getInstance(MESSAGE_DIGEST_ALGORITHM_SHA256,
+                                                    OPENJCE_PLUS_FIPS_NAME);
+                } else {
+                    md1 = MessageDigest.getInstance(MESSAGE_DIGEST_ALGORITHM_SHA256,
+                                                    IBMJCE_PLUS_FIPS_NAME);
+                }
+            } else {
```

Comment:
This is different than what was in `LTPAToken2` and `LTPADigSignature`.  Should there be an else if for IBM JCE available?

Reply:
If FIPS is enabled, we want to enforce it to use the right provider and fallback to the JCE provider

Reply:
You didn't address the question though.  Should there be an `else if` before the `else`.  Something like:

```suggestion
            } else if (LTPAKeyUtil.isIBMJCEAvailable()) {
                md1 = MessageDigest.getInstance(MESSAGE_DIGEST_ALGORITHM_SHA, IBMJCE_NAME);
            } else {
```
That is what used to be done before the `else` and now isn't done.  That would seem like a behavior change.

Reply:
SHA1 is available for IBMJCE and IBMJCEPLUS. If they don't have FIPS enabled, then they will get SHA1 ... any how better be safer!!! fixed it.